File: CPP/sorting/heapSort.cpp
#include <iostream>
#include <vector>
#include <algorithm> // For std::swap

// A helper function to heapify a subtree rooted with node i
// n is the size of the heap
void heapify(std::vector<int>& arr, int n, int i) {
    int largest = i;       // Initialize largest as root
    int left = 2 * i + 1;  // left child = 2*i + 1
    int right = 2 * i + 2; // right child = 2*i + 2

    // If left child is larger than root
    if (left < n && arr[left] > arr[largest]) {
        largest = left;
    }

    // If right child is larger than largest so far
    if (right < n && arr[right] > arr[largest]) {
        largest = right;
    }

    // If largest is not the root, swap them
    if (largest != i) {
        std::swap(arr[i], arr[largest]);

        // Recursively heapify the affected sub-tree
        heapify(arr, n, largest);
    }
}

// The main function to perform heap sort
void heapSort(std::vector<int>& arr) {
    int n = arr.size();

    // 1. Build a max-heap from the input data
    // We start from the last non-leaf node and move upwards
    for (int i = n / 2 - 1; i >= 0; i--) {
        heapify(arr, n, i);
    }

    // 2. Extract elements from heap one by one
    for (int i = n - 1; i > 0; i--) {
        // Move current root (max element) to the end
        std::swap(arr[0], arr[i]);

        // Call max heapify on the reduced heap
        heapify(arr, i, 0);
    }
}

// Utility function to print a vector
void printArray(const std::vector<int>& arr) {
    for (int i : arr) {
        std::cout << i << " ";
    }
    std::cout << std::endl;
}

// Main function to test the implementation
int main() {
    std::vector<int> arr = {12, 11, 13, 5, 6, 7};

    std::cout << "Original array: ";
    printArray(arr);

    heapSort(arr);

    std::cout << "Sorted array:   ";
    printArray(arr);

    return 0;
}

File: heap_sort.py
#!/usr/bin/env python3
"""Heap Sort - O(n log n)"""
def heapify(a,n,i):
    l,r=2*i+1,2*i+2; largest=i
    if l<n and a[l]>a[largest]: largest=l
    if r<n and a[r]>a[largest]: largest=r
    if largest!=i:
        a[i],a[largest]=a[largest],a[i]
        heapify(a,n,largest)
def heap_sort(a):
    n=len(a)
    for i in range(n//2-1,-1,-1): heapify(a,n,i)
    for i in range(n-1,0,-1):
        a[i],a[0]=a[0],a[i]; heapify(a,i,0)
    return a
if __name__=="__main__":
    arr=[4,10,3,5,1]
    print(heap_sort(arr))
